src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/ApiConventionItem.cs

```diff
@@ -0,0 +1,199 @@
+            {
+                var typeMatchAttribute = parameter.GetCustomAttribute<ApiConventionTypeMatchAttribute>(inherit: false);
+                return typeMatchAttribute?.MatchBehavior ?? ApiConventionTypeMatchBehavior.Exact;
+            }
+        }
+
+        internal static bool IsNameMatch(string name, string conventionName, ApiConventionNameMatchBehavior nameMatchBehavior)
+        {
+            switch (nameMatchBehavior)
+            {
+                case ApiConventionNameMatchBehavior.Any:
+                    return true;
+
+                case ApiConventionNameMatchBehavior.Exact:
+                    return string.Equals(name, conventionName, StringComparison.Ordinal);
+
+                case ApiConventionNameMatchBehavior.Prefix:
+                    return IsNameMatchPrefix();
+
+                case ApiConventionNameMatchBehavior.Suffix:
+                    return IsNameMatchSuffix();
+
+                default:
+                    return false;
+            }
+
+            bool IsNameMatchPrefix()
+            {
+                if (!name.StartsWith(conventionName, StringComparison.Ordinal))
+                {
+                    return false;
+                }
+
+                if (name.Length == conventionName.Length)
+                {
+                    return true;
+                }
+
+                return char.IsUpper(name[conventionName.Length]);
+            }
+
+            bool IsNameMatchSuffix()
+            {
+                // name = id, conventionName = id
+                if (string.Equals(name, conventionName, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+
+                if (name.Length <= conventionName.Length)
+                {
+                    return false;
+                }
+
+                // name = personId, conventionName = id
+                var index = name.Length - conventionName.Length - 1;
+                if (!char.IsLower(name[index]))
+                {
+                    return false;
+                }
+
+                index++;
+                if (name[index] != char.ToUpper(conventionName[0]))
+                {
+                    return false;
+                }
+
+                index++;
+                return string.Compare(name, index, conventionName, 1, conventionName.Length - 1, StringComparison.Ordinal) == 0;
+            }
+        }
+
+        internal static bool IsTypeMatch(Type type, Type conventionType, ApiConventionTypeMatchBehavior typeMatchBehavior)
+        {
+            switch (typeMatchBehavior)
+            {
+                case ApiConventionTypeMatchBehavior.Any:
+                    return true;
+
+                case ApiConventionTypeMatchBehavior.Exact:
+                    return type == conventionType;
+
+                case ApiConventionTypeMatchBehavior.AssignableFrom:
+                    return conventionType.IsAssignableFrom(type);
```

Comment:
I'm not sure that this distinction is important.

Reply:
As-in, I'm not sure we need `Exact`

Reply:
Ok. I can take that out.